modeling/algorithm/local-weighted-slice-algorithm.go

```diff
@@ -0,0 +1,202 @@
+		} else {
+			endpointsNeeded.push(endpointDeviation{name: zoneName, deviation: -intDeviation})
+			localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: zone.Endpoints, Weight: 1}
+		}
+		sliceGroups[zoneName] = localGroup
+
+		// push 0.xx deviation into corresponding lists
+		floatDeviation := deviation - float64(intDeviation)
+		if floatDeviation > 0 {
+			// Intentionally set deviation to default 0, this means the zone
+			// itself needs float part from this endpoint otherwise means the
+			// endpoints will be shared by other zones only
+			weightedEndpointsAvailable.push(endpointDeviation{name: zoneName, weight: floatDeviation})
+		} else if floatDeviation < 0 {
+			weightedEndpointsNeeded.push(endpointDeviation{name: zoneName, deviation: 1, weight: -floatDeviation})
+		}
+	}
+
+	err := alg.balanceSliceGroups(&endpointsAvailable, &endpointsNeeded, &weightedEndpointsAvailable, &weightedEndpointsNeeded, sliceGroups)
+	return sliceGroups, err
+}
+
+// balanceSliceGroups distributes endpoints from zones with extra endpoints to
+// EndpointSliceGroups for zones with insufficient endpoints.
+func (alg LocalWeightedSliceAlgorithm) balanceSliceGroups(endpointsAvailable *endpointsList, endpointsNeeded *endpointsList, weightedEndpointsAvailable *endpointsList, weightedEndpointsNeeded *endpointsList, sliceGroups map[string]types.EndpointSliceGroup) error {
+
+	for _, receiveZone := range endpointsNeeded.byZone {
+		// No more 'int' endpoints available but are still 'int' endpoints in
+		// need, we push the needed entpoints into weighted list and deal them
+		// as float endpoints
+		if len(endpointsAvailable.byZone) == 0 {
+			receiveZone.weight = 1
+			weightedEndpointsNeeded.push(receiveZone)
+			endpointsNeeded.pop()
+			continue
+		}
+		// same logic as original local-slice algorithm
+		for index := 0; index < len(endpointsAvailable.byZone); {
+			sendZone := endpointsAvailable.byZone[index]
+			if sendZone.deviation == receiveZone.deviation {
+				sliceGroups[receiveZone.name].Composition[sendZone.name] = types.WeightedEndpoints{Number: sendZone.deviation, Weight: 1}
+				receiveZone.deviation = 0
+				endpointsAvailable.pop()
+				break
+			}
+			if sendZone.deviation > receiveZone.deviation {
+				sliceGroups[receiveZone.name].Composition[sendZone.name] = types.WeightedEndpoints{Number: receiveZone.deviation, Weight: 1}
+				endpointsAvailable.byZone[index].deviation -= receiveZone.deviation
+				receiveZone.deviation = 0
+				break
+			}
+			if sendZone.deviation < receiveZone.deviation {
+				sliceGroups[receiveZone.name].Composition[sendZone.name] = types.WeightedEndpoints{Number: sendZone.deviation, Weight: 1}
+				receiveZone.deviation -= sendZone.deviation
+				endpointsAvailable.pop()
+				continue
+			}
+		}
+		// if needed.deviation != 0 means more 'int' needed than available, push
+		// to weighted list and deal as float
```

Comment:
Isn't this similar to the comment on 103?

Reply:
yes, but this is an edge case that the '103' can't deal with. Simply its one deviation has remaining number needed but no more available endpoints to serve it, we have to push that deviation into float list and treat as float endpoints.